softflowd/softflowctl.8
.\" $Id: softflowctl.8,v 1.1.1.1 2009/01/28 14:14:16 akrifa Exp $
.\"
.\" Copyright (c) 2002 Damien Miller.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd October 18, 2002
.Dt SOFTFLOWCTL 8
.Os
.Sh NAME
.Nm softflowctl
.Nd Remote control program for softflowd
.Sh SYNOPSIS
.Nm softflowctl
.Op Fl h
.Op Fl c Ar ctl_sock
.Ar command
.Sh DESCRIPTION
.Nm
is a remote control program used to control a running
.Xr softflowd 8 
daemon.
.Pp
The command-line options are as follows:
.Bl -tag -width Ds
.It Fl c Ar ctlsock
Specify an alternate location for the remote control socket.
Default is 
.Pa /var/run/softflowd.ctl
.It Fl h
Displays commandline usage information.
.El
.Pp
.Sh COMMANDS
.Bl -tag -width Ds
.It Pa shutdown
Ask
.Xr softflowd 8
to gracefully exit. This is equivalent to sending it a 
.Dv SIGTERM 
or 
.Dv SIGINT .
.It Pa exit
Ask
.Xr softflowd 8
to immediately exit. No flow expiry processing or data export is performed.
.It Pa expire-all
Immediately expire all tracked flows.
.It Pa delete-all
Immediately delete all tracked flows. No flow expiry processing or data 
export is performed.
.It Pa statistics
Return statistics collected by 
.Xr softflowd 8
on expired flows.
.It Pa debug+
Increase the debugging level of 
.Xr softflowd 8
.It Pa debug-
Decrease the debugging level.
.It Pa stop-gather
Stops network data collection by
.Xr softflowd 8 .
.It Pa start-gather
Resumes network data collection.
.It Pa dump-flows
Return information on all tracked flows.
.It Pa timeouts
Print information on flow timeout parameters.
.It Pa send-template
Resend a NetFlow v.9 template record before the next flow export.
Has no effect for other flow export versions.
.El
.Sh AUTHORS
Damien Miller <djm@mindrot.org>
.Sh SEE ALSO
.Xr softflowd 8


flowd/ChangeLog
20080725
 - (djm) byte swap interface indices before testing them in filters
 - (djm) Fixed swapped last/first_switched times in netflow v.9 code
   patch from weinhold AT berbee.com
 - (djm) Released flowd-0.9.1

20080422
 - (djm) Support flow-tools CSV output format in flowd-reader. Patch from
   weinhold AT berbee.com
 - (djm) Use proper API for logging debug information; Spotted by
   JSaxe AT briworks.com

20071024
 - (djm) Support explicit specification of "listen on" and "logsock"
   socket buffer sizes. Patch from kempfj2 AT cs.rpi.edu

20071010
 - (djm) Make local socket-only logging work, and fix spurious warnings
   from socket logging. Patch from kempfj2 AT cs.rpi.edu
 - (djm) Unbreak flowinsert.pl

20070723
 - (djm) Fix NetFlow v.9 flowset template parsing on LP!32 platforms.
   Report and fix from zhangjinxue AT cernet.edu.cn

20070510
 - (djm) Add support for filtering input/output interface index. Requested
   by Ralf Kleineisel, debugging assistance from Matthew Smart

20061020
 - (djm) Rename CRC32 functions to avoid collisions with zlib;
   patch from Sergey Vasilenko <stalker@telemost.ru>

20051227
 - (djm) Add RPM spec and init files for SuSE Linux from alshu AT tut.by
 - (djm) Released flowd-0.9

20051220
 - (djm) Fix byte swapping of src/dst AS and interface indices, spotted 
   and fix tested by Gijs Molenaar

20051208
 - (djm) Implement absolute time (before & after) filters, requested by
   alshu AT tut.by

20051207
 - (djm) Demote input buffer full warning to a debug and reduce the length
   of the input queue (now that we have an output queue)
 - (djm) Another hex that got away

20051206
 - (djm) Rename some filter members to make absolute time filtering easier

20051204
 - (djm) Add an output queue, so we don't do tiny little filesystem writes for
   each flow we receive
 - (djm) Sync tree.h and queue.h with OpenBSD

20051117
 - (djm) Prefix all hex integers in logs with "0x"
 - (djm) More source_id debugging everywhere

20051111
 - (djm) Support devices that send multiple templates in a single template
   packet section. Thanks to Gijs Molenaar for packet dumps that demonstrated
   this
 - (djm) Turn off Netflow v.9 debugging

20051013
 - (djm) Fix bug that broke filtering on address family in flows, spotted by
   Gijs Molenaar
 - (djm) Rename common.h -> flowd-common.h, always install it and
   flowd-config.h. Should help platforms without intXX_t etc.

20051001
 - (djm) Add FlowLog_from_file method to Python module
 - (djm) Increase UDP socket receive buffer size, shrink socket send buffer size
   to cope better with bursts of flows
 - (djm) Allow reading from standard input in tools/stats.py
 - (djm) Better error message on corrupt flows
 - (djm) Change flowd-reader to *overwrite* existing log files, not *append* to
   them

20050928
 - (djm) Add a basic input queue to flowd, to improve its behaviour when it
   receives sudden bursts of packets.

20050918
 - (djm) Fix error reporting on logsock send failures
 - (djm) Fix another typo in tools/stats.py, spotted by alshu@tut.by
 - (djm) Sync usage for flowd-reader
 - (djm) -Wall cleanup flowd_python.c
 - (djm) Tidy stats.py, and remove hard dependency on curses module

20050918
 - (djm) Fix typo in tools/stats.py, spotted by alshu@tut.by
 - (djm) Mention softflowd and pfflowd in README, suggested by alshu@tut.by

20050913
 - (djm) Mention creation of _flowd user in INSTALL, spotted by alshu@tut.by

20050826
 - (djm) Add interval_time() and iso_time() functions to Python module
 - (djm) Port and tidy stats script
 - (djm) Add a "head" mode to flowd-reader

20050826
 - (djm) Add Flow.has_field() to Python API

20050825
 - (djm) Make it compile on Solaris 9 (Perl module is still busted because 
   of compiler misguessing stupidity)

20050824
 - (djm) Add support for relaying serialised flows to a local Unix domain
   datagram socket in realtime. This option ("logsock" in flowd.conf) is
   considered experimental. A sample Python client exists in 
   tools/sockclient.py that receives and prints the flows.

20050822
 - (djm) Improve the Python API some more:
   - Prefer PyLong as type for various fields
   - Maintain all addresses as PyObjects rather than char*, so it is 
     possible to write to them
   - Initialise missing addresses, octet and packet counters to Py_None
   - Fix lots of bugs
 - (djm) Macroise much of the NetFlow v.9 parsing code and fix length
   mismatches in the process
 - (djm) Check for, and refuse to append to legacy logfiles
 - (djm) Don't lookup username when running in INSECURE mode
 
20050822
 - (djm) Figure out Python structmember.h types for struct store_flow_complete
   at autoconf time and stuff them into their own header
 - (djm) Add (src|dst|agent|gateway)_addr_af to Python API

20050821
 - (djm) Major rewrite of storage code:
   - Introduce a new storage format (version 3), which should be much faster
     to read from disk and easier and more graceful to extend
   - Add FILE* oriented API
   - Add some new fields: receive time microseconds, netflow v.9 source_id
   - Extend the widths of several fields: if_ndx_(in|out), (src|dst)_as, 
     engine_(type|id)
   - Add functions for reading and writing legacy logs
 - (djm) Major rewrite of Python API
   - Implements all the functions in all hybrid API using 100% C code
   - Use Python structmember API and store.c for major speed improvement
   - Adds an iterator object to read all flows from a flow log
 - (djm) Adjust Perl API to cope with new store format
 - (djm) Add ability for flowd-reader to read and convert legacy log files
 - (djm) Update copyright years

20050707
 - (djm) Fix pidfile path in manpage

20050619
 - (djm) README typo, spotted by cruel AT texnika.com.ua

200500601
 - (djm) Add a simple Python summarisation and charting script (tools/stats.py)

20050514
 - (djm) Fix spurious error message
 - (djm) Improve day filters, allowing lists and ranges of days
 - (djm) Release flowd-0.8.5

20050428

softflowd/softflowctl.c
/*
 * Copyright 2002 Damien Miller <djm@mindrot.org> All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "common.h"

RCSID("$Id: softflowctl.c,v 1.1.1.1 2009/01/28 14:14:16 akrifa Exp $");

static void
usage(void)
{
	fprintf(stderr, "Usage: [-c ctlsock] softflowctl [command]\n");
}

int
main(int argc, char **argv)
{
	const char *ctlsock_path;
	char buf[8192], *command;
	struct sockaddr_un ctl;
	socklen_t ctllen;
	int ctlsock, ch;
	FILE *ctlf;
	extern char *optarg;
	extern int optind;

	ctlsock_path = DEFAULT_CTLSOCK;
	while ((ch = getopt(argc, argv, "hc:")) != -1) {
		switch (ch) {
		case 'h':
			usage();
			return (0);
		case 'c':
			ctlsock_path = optarg;
			break;
		default:
			fprintf(stderr, "Invalid commandline option.\n");
			usage();
			exit(1);
		}
	}
	
	/* Accept only one argument */
	if (optind != argc - 1) {
		usage();
		exit(1);
	}
	command = argv[optind];

	memset(&ctl, '\0', sizeof(ctl));
	if (strlcpy(ctl.sun_path, ctlsock_path, sizeof(ctl.sun_path)) >= 
	    sizeof(ctl.sun_path)) {
		fprintf(stderr, "Control socket path too long.\n");
		exit(1);
	}
	ctl.sun_path[sizeof(ctl.sun_path) - 1] = '\0';
	ctl.sun_family = AF_UNIX;
	ctllen = offsetof(struct sockaddr_un, sun_path) +
            strlen(ctlsock_path) + 1;
#ifdef SOCK_HAS_LEN 
	ctl.sun_len = ctllen;
#endif
	if ((ctlsock = socket(PF_UNIX, SOCK_STREAM, 0)) < 0) {
		fprintf(stderr, "ctl socket() error: %s\n", 
		    strerror(errno));
		exit(1);
	}
	if (connect(ctlsock, (struct sockaddr*)&ctl, sizeof(ctl)) == -1) {
		fprintf(stderr, "ctl connect(\"%s\") error: %s\n",
		    ctl.sun_path, strerror(errno));
		exit(1);
	}
	
	if ((ctlf = fdopen(ctlsock, "r+")) == NULL) {
		fprintf(stderr, "fdopen: %s\n", strerror(errno));
		exit(1);
	}
	setlinebuf(ctlf);
	
	/* Send command */
	if (fprintf(ctlf, "%s\n", command) < 0) {
		fprintf(stderr, "write: %s\n", strerror(errno));
		exit(1);
	}

	/* Write out reply */
	while((fgets(buf, sizeof(buf), ctlf)) != NULL)
		fputs(buf, stdout);

	fclose(ctlf);
	close(ctlsock);

	exit(0);
}
